File: core/src/com/xenoage/zong/commands/core/music/group/BracketGroupAdd.java
package com.xenoage.zong.commands.core.music.group;

import java.util.List;

import com.xenoage.utils.document.command.Command;
import com.xenoage.utils.document.command.Undoability;
import com.xenoage.zong.core.music.StavesList;
import com.xenoage.zong.core.music.group.BarlineGroup;
import com.xenoage.zong.core.music.group.BracketGroup;


/**
 * Adds a {@link BarlineGroup} to the score.
 * 
 * @author Andreas Wenger
 */
public class BracketGroupAdd
	implements Command {

	//data
	private StavesList stavesList;
	private BracketGroup group;


	public BracketGroupAdd(StavesList stavesList, BracketGroup group) {
		this.stavesList = stavesList;
		this.group = group;
	}


	@Override public void execute() {
		//check parameters
		int startStaff = group.getStaves().getStart();
		int endStaff = group.getStaves().getStop();
		if (endStaff >= stavesList.getStaves().size())
			throw new IllegalArgumentException("staves out of range");

		//add new group at the right position
		//(the barline groups are sorted by start index)
		List<BracketGroup> groups = stavesList.getBracketGroups();
		int i = 0;
		while (i < groups.size() && startStaff > groups.get(i).getStaves().getStart())
			i++;
		groups.add(i, group);
	}


	@Override public Undoability getUndoability() {
		return Undoability.Undoable;
	}


	@Override public void undo() {
		stavesList.getBracketGroups().remove(group);
	}

}

File: core/src/com/xenoage/zong/commands/core/music/group/BracketGroupAdd.kt
package com.xenoage.zong.commands.core.music.group

import com.xenoage.utils.document.command.UndoableCommand
import com.xenoage.zong.core.music.StavesList
import com.xenoage.zong.core.music.group.BarlineGroup
import com.xenoage.zong.core.music.group.BracketGroup


/**
 * Adds a [BarlineGroup] to the score.
 */
class BracketGroupAdd(
		private val stavesList: StavesList,
		private val group: BracketGroup
) : UndoableCommand() {

	override fun execute() {
		//check parameters
		val startStaff = group.staves.start
		val endStaff = group.staves.stop
		if (endStaff >= stavesList.staves.size)
			throw IllegalArgumentException("staves out of range")

		//add new group at the right position
		//(the barline groups are sorted by start index)
		val groups = stavesList.bracketGroups
		var i = 0
		while (i < groups.size && startStaff > groups[i].staves.start)
			i++
		groups.add(i, group)
	}

	override fun undo() {
		stavesList.bracketGroups.remove(group)
	}

}
